partiql-app/partiql-cli/src/main/kotlin/org/partiql/cli/utils/DelimitedFileIterable.kt

```diff
@@ -0,0 +1,35 @@
+package org.partiql.cli.utils
+
+import com.amazon.ion.IonSystem
+import org.apache.commons.csv.CSVFormat
+import org.partiql.cli.functions.ReadFile.Companion.conversionModeFor
+import org.partiql.lang.eval.ExprValue
+import org.partiql.lang.eval.io.DelimitedValues
+import java.io.InputStreamReader
+
+/**
+ * A closeable, iterable wrapper over an [InputSource] containing Ion values. With an [DelimitedFileIterable],
+ * it is possible to create a new [Iterator] of [ExprValue]'s from an [InputSource] containing a sequence of delimited
+ * values. By leveraging [InputSource]'s ability to re-open streams, the [DelimitedFileIterable] is able to create
+ * iterators at the beginning of the [InputSource.stream].
+ */
+internal class DelimitedFileIterable(
+    private val ion: IonSystem,
+    private val input: InputSource,
+    private val format: CSVFormat,
+    private val encoding: String,
+    private val conversion: String
+) : Iterable<ExprValue>, AutoCloseable {
+
+    private val readers: MutableList<InputStreamReader> = mutableListOf()
+
+    override fun iterator(): Iterator<ExprValue> {
+        val reader = InputStreamReader(input.stream(), encoding)
+        readers.add(reader)
+        return DelimitedValues.exprValue(ion, reader, format, conversionModeFor(conversion)).iterator()
+    }
+
+    override fun close() {
+        readers.forEach { it.close() }
+    }
+}
```

Comment:
I am not sure what is going on here.  If I am confused, please help me understand.

The point of `DelimitedValues.exprValue` is to create a lazy sequence, while keeping the `reader` open.  In other words, even though the sequence is "created", it is not traversed until someone actually traverses it.  Until that traversal happens, the `reader` must be kept open.  
Now, the only call site of this iterator method is this, in ReadFile:
```
        val iterable = DelimitedFileIterable(ion, inputSource, csvFormatWithOptions, encoding, conversion)
        iterable.use { ExprValue.newBag(it) }
```
and it does not do the traversal, from what I understand: 
The call to `ExprValue.newBag` receives `it`, an instance of `DelimitedFileIterable` from the previous line and calls its `iterator` method from Line 26 above, which opens an iterator on the reader.  The only thing that is done after this by the code block passed to `use` is wrapping the iterator as an `ExprValue` (`BagExprValue` specifically).  Importantly, nothing is yet read from the reader.  But immediately after that, the 'use' method initiates closing the reader, through the method `close` on Line 32.  

Separately, why is the complication with keeping the list of readers?  In the above (and only) call site, the `iterator` method seems to be called only once and, therefore, only one reader is created. So, whatever the complication is for, is it being exercised?  But if it were, wouldn't `close()` close all the readers at once, regardless of whether they all have been already consumed?